File: ruby/lib/cck/examples.rb
# frozen_string_literal: true

module CCK
  module Examples
    class << self
      def gherkin
        Dir.entries(cck_features_folder_location).select do |file_or_folder|
          next if file_or_folder.start_with?('.')

          gherkin_example?(File.join(cck_features_folder_location, file_or_folder))
        end
      end

      def markdown
        Dir.entries(cck_features_folder_location).select do |file_or_folder|
          next if file_or_folder.start_with?('.')

          markdown_example?(File.join(cck_features_folder_location, file_or_folder))
        end
      end

      def feature_code_for(example_name)
        path = File.join(cck_features_folder_location, example_name)

        return path if File.directory?(path)

        raise ArgumentError, "No feature code directory found in gem for CCK example: #{example_name}"
      end

      private

      def cck_features_folder_location
        File.expand_path("#{File.dirname(__FILE__)}/../../features/")
      end

      def gherkin_example?(example_folder)
        file_type_in_folder?('.feature', example_folder)
      end

      def markdown_example?(example_folder)
        file_type_in_folder?('.md', example_folder)
      end

      def file_type_in_folder?(extension, folder)
        Dir.entries(folder).any? { |file| File.extname(file) == extension }
      end
    end
  end
end

File: python/src/cck/examples.py
from pathlib import Path


class Examples:
    def __init__(self):
        self.cck_features_folder_location = Path(__file__).resolve().parent.parent.parent / 'features'

    def gherkin(self):
        """
        Return the paths for each example that is of 'gherkin' type.
        """
        return [
            folder for folder in self.cck_features_folder_location.iterdir()
            if not folder.name.startswith('.') and
            self.gherkin_example(folder)
        ]

    def markdown(self):
        """
        Return the paths for each example that is of 'markdown' type.
        """
        return [
            folder for folder in self.cck_features_folder_location.iterdir()
            if not folder.name.startswith('.') and
            self.markdown_example(folder)
        ]

    def feature_code_for(self, example_name):
        """
        Return the path for a specific example scenario based on its name.
        """
        path = self.cck_features_folder_location / example_name
        if path.is_dir():
            return path
        raise ValueError(f"No feature code directory found for CCK example: {example_name}")

    def gherkin_example(self,example_folder):
        """
        Check if the example folder contains a '.feature' file.
        """
        return self.file_type_in_folder('.feature', example_folder)

    def markdown_example(self, example_folder):
        """
        Check if the example folder contains a '.md' file.
        """
        return self.file_type_in_folder('.md', example_folder)

    @staticmethod
    def file_type_in_folder(extension, folder):
        """
        Check if the folder contains any file with the given extension.
        """
        return any(file.suffix == extension for file in folder.iterdir())
